File: Darts/src/burlton/dartzee/code/object/Dart.java
package burlton.dartzee.code.object;

import java.awt.Point;
import java.util.ArrayList;

import burlton.dartzee.code.db.GameEntity;
import burlton.dartzee.code.utils.DartsRegistry;
import burlton.dartzee.code.utils.PreferenceUtil;
import burlton.core.code.util.StringUtil;

public class Dart implements DartsRegistry
{
	private Point pt = null;
	private int score = -1;
	private int multiplier = -1; //double or triple
	private int segmentType = -1; //Outer single, miss, etc
	
	private int ordinal = -1;
	
	//What the player's "score" was before throwing this dart.
	//Means what you'd think for X01
	//For Round the Clock, this'll be what they were to aim for
	private int startingScore = 1; 
	
	//Never set on the DB. Used for in-game stats, and is just set to the round number.
	private int golfHole = -1;
	private long participantId = -1;
	
	public Dart(int score, int multiplier)
	{
		this.score = score;
		this.multiplier = multiplier;
	}
	public Dart(Point pt, int score, int multiplier, int segmentType)
	{
		this.pt = pt;
		this.score = score;
		this.multiplier = multiplier;
		this.segmentType = segmentType;
	}
	
	/**
	 * Helpers
	 */
	public boolean isDouble()
	{
		return multiplier == 2;
	}
	public boolean isTriple()
	{
		return multiplier == 3;
	}
	public int getTotal()
	{
		return score * multiplier;
	}
	public int getGolfScore(int target)
	{
		if (score != target)
		{
			return 5;
		}
		
		return DartboardSegment.getGolfScore(segmentType);
	}
	
	public int getGolfScore()
	{
		return getGolfScore(golfHole);
	}
	
	
	/**
	 * Gets / sets
	 */
	public int getScore() 
	{
		return score;
	}
	public void setScore(int score) 
	{
		this.score = score;
	}
	public int getMultiplier() 
	{
		return multiplier;
	}
	public void setMultiplier(int multiplier) 
	{
		this.multiplier = multiplier;
	}
	public int getOrdinal()
	{
		return ordinal;
	}
	public void setOrdinal(int ordinal)
	{
		this.ordinal = ordinal;
	}
	public int getX()
	{
		return (int)pt.getX();
	}
	public int getY()
	{
		return (int)pt.getY();
	}
	public int getSegmentType()
	{
		return segmentType;
	}
	public void setSegmentType(int segmentType)
	{
		this.segmentType = segmentType;
	}
	public int getStartingScore()
	{
		return startingScore;
	}
	public void setStartingScore(int clockNumber)
	{
		this.startingScore = clockNumber;
	}
	public void setGolfHole(int hole)
	{
		this.golfHole = hole;
	}
	public long getParticipantId()
	{
		return participantId;
	}
	public void setParticipantId(long participantId)
	{
		this.participantId = participantId;
	}
	
	@Override
	public int hashCode()
	{
		final int prime = 31;
		int result = 1;
		result = prime * result + multiplier;
		result = prime * result + ordinal;
		result = prime * result + score;
		return result;
	}

	@Override
	public boolean equals(Object obj)
	{
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (!(obj instanceof Dart))
			return false;
		Dart other = (Dart) obj;
		if (multiplier != other.multiplier)
			return false;
		if (ordinal != other.ordinal)
			return false;
		if (score != other.score)
			return false;
		return true;
	}

	@Override
	public String toString() 
	{
		boolean showTotal = PreferenceUtil.getBooleanValue(PREFERENCES_BOOLEAN_DISPLAY_DART_TOTAL_SCORE);
		if (showTotal)
		{
			return "" + getTotal();
		}
		
		return getRendered();
	}
	
	public boolean hitClockTarget(String clockType)
	{
		if (clockType.equals(GameEntity.CLOCK_TYPE_DOUBLES)
		  && !isDouble())
		{
			return false;
		}
		
		if (clockType.equals(GameEntity.CLOCK_TYPE_TREBLES)
		  && !isTriple())
		{
			return false;
		}
		
		return score == startingScore
		  && multiplier > 0;
	}
	
	public String getRendered()
	{
		if (multiplier == 0)
		{
			return "0";
		}
		
		String ret = "";
		if (isDouble())
		{
			ret += "D";
		}
		else if (isTriple())
		{
			ret += "T";
		}
		
		ret += score;
		
		return ret;
	}
	
	public int getSegmentTypeToAimAt()
	{
		if (multiplier == 1)
		{
			return DartboardSegment.TYPE_OUTER_SINGLE;
		}
		else if (multiplier == 2)
		{
			return DartboardSegment.TYPE_DOUBLE;
		}
		
		return DartboardSegment.TYPE_TREBLE;
	}
	
	/**
	 * Static methods
	 */
	public static Dart factorySingle(int score)
	{
		return new Dart(score, 1);
	}
	public static Dart factoryDouble(int score)
	{
		return new Dart(score, 2);
	}
	public static Dart factoryTreble(int score)
	{
		return new Dart(score, 3);
	}
	
	public int compareTo(Dart other)
	{
		//If there's a strict inequality in total, then it's simple
		if (getTotal() > other.getTotal())
		{
			return 1;
		}
		
		if (getTotal() < other.getTotal())
		{
			return -1;
		}
		
		//Totals are equal. So now look at the multiplier.
		//I.e. T12 > D18 even though the totals are both 36.
		if (getMultiplier() > other.getMultiplier())
		{
			return 1;
		}
		
		if (other.getMultiplier() > getMultiplier())
		{
			return -1;
		}
		
		//Same total and same multiplier, so must be an equivalent dart
		return 0;
	}
}

File: Darts/src/burlton/dartzee/code/object/Dart.kt
package burlton.dartzee.code.`object`

import burlton.dartzee.code.db.CLOCK_TYPE_DOUBLES
import burlton.dartzee.code.db.CLOCK_TYPE_TREBLES
import burlton.dartzee.code.utils.PREFERENCES_BOOLEAN_DISPLAY_DART_TOTAL_SCORE
import burlton.dartzee.code.utils.PreferenceUtil
import burlton.desktopcore.code.util.DateStatics
import java.awt.Point
import java.sql.Timestamp

open class Dart constructor(
        var score: Int,
        var multiplier: Int,
        var pt: Point? = null,
        var segmentType: Int = -1)
{
    var ordinal = -1

    //What the player's "score" was before throwing this dart.
    //Means what you'd think for X01
    //For Round the Clock, this'll be what they were to aim for
    var startingScore = -1

    //Never set on the DB. Used for in-game stats, and is just set to the round number.
    private var golfHole = -1
    var participantId: String = ""
    var gameId: String = ""
    var dtThrown: Timestamp = DateStatics.END_OF_TIME

    /**
     * Helpers
     */
    fun isDouble() = multiplier == 2
    fun isTriple() = multiplier == 3
    fun getTotal() = score * multiplier
    fun getGolfScore() = getGolfScore(golfHole)
    fun getX() = pt?.x
    fun getY() = pt?.y

    fun getHitScore(): Int
    {
        return when(multiplier)
        {
            0 -> 0
            else -> score
        }
    }

    fun getRendered(): String
    {
        if (multiplier == 0)
        {
            return "0"
        }

        var ret = ""
        if (isDouble())
        {
            ret += "D"
        }
        else if (isTriple())
        {
            ret += "T"
        }

        ret += score

        return ret
    }

    fun getSegmentTypeToAimAt(): Int
    {
        return when (multiplier)
        {
            1 -> SEGMENT_TYPE_OUTER_SINGLE
            2 -> SEGMENT_TYPE_DOUBLE
            else -> SEGMENT_TYPE_TREBLE
        }
    }

    fun getGolfScore(target: Int): Int
    {
        return if (score != target)
        {
            5
        } else getGolfScoreForSegment(segmentType)

    }

    fun setGolfHole(hole: Int)
    {
        this.golfHole = hole
    }

    override fun hashCode(): Int
    {
        val prime = 31
        var result = 1
        result = prime * result + multiplier
        result = prime * result + ordinal
        result = prime * result + score
        return result
    }

    override fun equals(other: Any?): Boolean
    {
        if (this === other)
            return true
        if (other == null)
            return false
        if (other !is Dart)
            return false
        if (multiplier != other.multiplier)
            return false
        if (ordinal != other.ordinal)
            return false
        return (score == other.score)
    }

    override fun toString(): String
    {
        val showTotal = PreferenceUtil.getBooleanValue(PREFERENCES_BOOLEAN_DISPLAY_DART_TOTAL_SCORE)
        return if (showTotal)
        {
            "" + getTotal()
        }
        else getRendered()

    }

    fun hitClockTarget(clockType: String?): Boolean
    {
        return when (clockType)
        {
            CLOCK_TYPE_DOUBLES -> score == startingScore && isDouble()
            CLOCK_TYPE_TREBLES -> score == startingScore && isTriple()
            else -> score == startingScore && multiplier > 0
        }
    }

    operator fun compareTo(other: Dart): Int
    {
        //If there's a strict inequality in total, then it's simple
        if (getTotal() > other.getTotal())
        {
            return 1
        }

        if (getTotal() < other.getTotal())
        {
            return -1
        }

        //Totals are equal. So now look at the multiplier.
        //I.e. T12 > D18 even though the totals are both 36.
        if (multiplier > other.multiplier)
        {
            return 1
        }

        return if (other.multiplier > multiplier)
        {
            -1
        }
        //Same total and same multiplier, so must be an equivalent dart
        else 0
    }
}

class DartNotThrown : Dart(-1, -1)

/**
 * Static methods
 */
fun factorySingle(score: Int): Dart
{
    return Dart(score, 1)
}

fun factoryDouble(score: Int): Dart
{
    return Dart(score, 2)
}

fun factoryTreble(score: Int): Dart
{
    return Dart(score, 3)
}